### DISCUSSION TITLE
"Keygen for v2.1s" doesn't work

### DISCUSSION THREAD

**Discussion by jopl** [deividAlfa/stm32_soldering_iron_controller#105] [Q&A] [answered]
Dear T12 enthusiastics,
I have just tested various firmwares and I found out the codes RG1 & RG2 created by gen.html don't work. Is there some mysterious step, which I am not doing to finish testing the KSGER FW, please? Or could it be possible the MCU STM32F103CBT6 uses another type of algorithm to calculate codes, please?
Thank you for your reply in advance.
Kind regards, Josef.

**Comment by deividAlfa** [comment]
It's just for 2.1s. Not for combo or hot air.
Anyways, [c00lhax0r](https://radiokot.ru/forum/viewtopic.php?f=2&t=154278&start=4140) released a new keygen in Python, I added it to the [Readme](https://github.com/deividAlfa/stm32_soldering_iron_controller#backing-up-the-original-firmware), try it out.
Though it's in Russian, I asked the guy to translate it.

**Comment by jopl** [comment]
Hi deivid,

thank you very much for your reply.

Unfortunately the new python keygen code is also not working in my case ... I tried all 3 variants of code (i.e. T12 / combo / ... BTW I guess I don't have combo, I think, ... I used file "...\stm32_soldering_iron_controller-master\Original_FW\KSGER\Encoder mod\[HW2.1S]_[1.3OLED].hex" for my own design (in attachment ... if you are interested, I can send you the KiCad files).

Don't you know, where I can do a mistake, please?

Your firmware seems to work like a charm ... most probably I will stay with it, but I just would like to recognize the difference, you know ...

Thank you so much in advance for your replay again.

Josef 
______________________________________________________________
> Od: "deividAlfa" ***@***.***>
> Komu: "deividAlfa/stm32_soldering_iron_controller" ***@***.***>
> Datum: 02.02.2024 20:18
> Předmět: Re: [deividAlfa/stm32_soldering_iron_controller] "Keygen for v2.1s" doesn't work (Discussion #105)
>
 
It's just for 2.1s. Not for combo or hot air.
 Anyways, c00lhax0r <https://radiokot.ru/forum/viewtopic.php?f=2&t=154278&start=4140> released a new keygen in Python, I added it to the Readme <https://github.com/deividAlfa/stm32_soldering_iron_controller#backing-up-the-original-firmware>, try it out.
 Though it's in Russian, I asked the guy to translate it.

**Comment by jopl** [comment]
![T12_Driver_v21S](https://github.com/deividAlfa/stm32_soldering_iron_controller/assets/401876/683b310d-7418-45a7-9bb6-cc8c4a000ff3)
PCB design ;-)

**Reply by deividAlfa** [reply]
The Keygen depends only on the stm32 Unique ID.
Try other firmwares, like 2.11, 2.12 or Handskit.
If not working, I can't help, I don't own a ksger and know little about their firmware, you might ask in the forums.

**Comment by jopl** [comment]
Dear deivid,
thank you really very much for your project, help and explanation. I am fully satisfied with customized (your) firmware, I can say. It is amazing work.
Kind regards, Josef.
 
______________________________________________________________
> Od: "deividAlfa" ***@***.***>
> Komu: "deividAlfa/stm32_soldering_iron_controller" ***@***.***>
> Datum: 02.02.2024 23:56
> Předmět: Re: [deividAlfa/stm32_soldering_iron_controller] "Keygen for v2.1s" doesn't work (Discussion #105)
>
 
The Keygen depends only on the stm32 Unique ID.
 If not working, I can't help, I don't own a ksger.

**Comment by jopl** [comment, marked as answer]
Finally I found the reason ... the EEPROM has to be 24C08 (tested) or 24C16 (not tested). Otherwise the firmware is not able to communicate with the EEPROM chip. Despite the fact I am not using the original firmware (custom FW is more stable and can regulate temperature more accurately - thanks to David), I was trying to uncover the issue. From EEPROM 24C32 (which I had populated on the board) , there is change in communication protocol.
regards, Josef.
